app/Http/Controllers/API/UserController.php

```diff
@@ -0,0 +1,273 @@
+        // Process the paramters
+        //
+
+        // Endorsements
+        if ($paramIncludeEndorsements) {
+            foreach ($returnUsers as $user) {
+                $user->endorsements = $this->mapEndorsements($user->endorsements->where('expired', false)->where('revoked', false));
+            }
+        }
+
+        // Roles
+        if ($paramIncludeRoles) {
+            foreach ($returnUsers as $user) {
+                $user->roles = collect();
+
+                foreach (Area::all() as $area) {
+                    $areaRoles = $user->groups->where('pivot.area_id', $area->id)->pluck('name');
+
+                    if ($areaRoles->count()) {
+                        $user->roles[$area->name] = ($user->groups->where('pivot.area_id', $area->id)->pluck('name'));
+                    } else {
+                        $user->roles[$area->name] = null;
+                    }
+                }
+            }
+        }
+
+        // Trainings
+        if ($paramIncludeTraining) {
+            foreach ($returnUsers as $user) {
+                $user->training = $this->mapTrainings($user->trainings->where('status', '>=', 0));
+            }
+        }
+
+        //
+        // Return the final result
+        //
+        $returnUsers = $this->mapUsers($returnUsers, $paramIncludeName, $paramIncludeEmail, $paramIncludeDivisions, $paramIncludeEndorsements, $paramIncludeRoles, $paramIncludeTraining);
+        $returnUsers = $returnUsers->sortBy('id')->values();
+
+        return response()->json(['data' => [
+            $returnUsers,
+        ]], 200);
+    }
+
+    /**
+     * Map out the required user data
+     *
+     * @return Collection
+     */
+    private function mapUsers(Collection $users, bool $includeName, bool $includeEmail, bool $includeDivisions, bool $includeEndorsements, bool $includeRoles, bool $includeTraining)
+    {
+        return $users->map(function ($user) use ($includeName, $includeEmail, $includeDivisions, $includeEndorsements, $includeRoles, $includeTraining) {
+
+            $returnData = [];
+
+            $returnData['id'] = $user->id;
+            ($includeEmail) ? $returnData['email'] = $user->email : null;
+            ($includeName) ? $returnData['first_name'] = $user->first_name : null;
+            ($includeName) ? $returnData['last_name'] = $user->last_name : null;
+            $returnData['rating'] = $user->rating_short;
+            ($includeDivisions) ? $returnData['region'] = $user->region : null;
+            ($includeDivisions) ? $returnData['division'] = $user->division : null;
+            ($includeDivisions) ? $returnData['subdivision'] = $user->subdivision : null;
+            $returnData['atc_active'] = boolval($user->atc_active);
+            ($includeEndorsements) ? $returnData['endorsements'] = $user->endorsements : null;
+            ($includeRoles) ? $returnData['roles'] = $user->roles : null;
+            ($includeTraining) ? $returnData['training'] = $user->training : null;
```

Comment:
question: There's a possibility I'm missing something here. If the parameters to the endpoint specify which fields to populate, why should this function need to know that? It can map the fields, have them all start as null, and merge in the whole `returnData` object. Why check if it should be included when the field(s) in question wouldn't exist to begin with if they hadn't already been requested?

Reply:
With the danger of me missing your point, the idea of doing it the way it's done here is because the user model returns a lot more fields than we want to expose, e.g. OAuth tokens, timestamps of different sorts, user's settings etc. By explicitly defining fields, we only include the fields we want in the return response.

Let me know if that didn't answer the question.